### PR TITLE
Cloneop

### PR DESCRIPTION
**Pull request by luitjens**
Adds clone as an operator.  This allows convolution to work on operators.

### CODE REVIEW

File: include/matx_conv.h [hunk 1]
```diff
@@ -165,25 +166,59 @@ inline void conv1d_impl(OutputType &o, const In1Type &i1, const In2Type &i2,
...
 inline void conv1d(OutputType &o, const In1Type &i1, const In2Type &i2,
                    matxConvCorrMode_t mode, cudaStream_t stream) {
-  if constexpr ( In2Type::Rank()==1 && In1Type::Rank() >  In2Type::Rank() ) {
-    //broadcast path.  clone In2 across entire batch
+  if constexpr ( In1Type::Rank() >  In2Type::Rank() ) {
+    // broadcast i2 path.  clone i2 across batches
     
-    const int Rank = In1Type::Rank();
-    typename In2Type::shape_type shape[Rank];
+    const int LRank = In1Type::Rank();
```
Comment (cliffburdick) [inline include/matx_conv.h:172]: could be constexpr to be explicit

File: include/matx_conv.h [hunk 2]
```diff
@@ -165,25 +166,59 @@ inline void conv1d_impl(OutputType &o, const In1Type &i1, const In2Type &i2,
...
+    const int SRank = In2Type::Rank();
+    const int DRank = LRank - SRank;
```
Comment (cliffburdick) [inline include/matx_conv.h:174]: same

File: include/matx_conv.h [hunk 3]
```diff
@@ -165,25 +166,59 @@ inline void conv1d_impl(OutputType &o, const In1Type &i1, const In2Type &i2,
...
+    auto ci2 = clone<LRank>(i2, shape);
 
     static_assert(i1.Rank() == ci2.Rank());
 
     conv1d_impl(o, i1, ci2, mode, stream);
     	  
+  }  else if constexpr ( In2Type::Rank() >  In1Type::Rank()) {
+    // broadcast i1 path.  clone i1 across batches
+    
+    const int LRank = In2Type::Rank();
```
Comment (cliffburdick) [inline include/matx_conv.h:199]: constexpr

File: include/matx_conv.h [hunk 4]
```diff
@@ -165,25 +166,59 @@ inline void conv1d_impl(OutputType &o, const In1Type &i1, const In2Type &i2,
...
+    index_t shape[LRank];
+
+    // copy left-most dimensions from i2
+    #pragma unroll
+    for(int i = 0; i < DRank; i++) {
+      shape[i] = i2.Size(i);
+    }
+    
+    // set right most dimensions as matxKeepDim
+    for(int i = 0; i < SRank; i++) {
```
Comment (cliffburdick) [inline include/matx_conv.h:211]: unroll?

File: include/matx_tensor_ops.h [hunk 5]
```diff
@@ -565,6 +565,84 @@ __MATX_INLINE__
...
+        }
+        MATX_ASSERT(d == T::Rank(), matxInvalidDim);
+
+      };
+
+      template <typename... Is>
+        __MATX_INLINE__ __MATX_DEVICE__ __MATX_HOST__ auto operator()(Is... indices) const 
+        {
+
+          // convert variadic type to tupple so we can read/update
```
Comment (cliffburdick) [inline include/matx_tensor_ops.h:601]: tuple

**Review by cliffburdick — APPROVED** [review]

File: include/matx_conv.h (hunk 3 above)
Comment (luitjens) [reply include/matx_conv.h]: fixed

File: include/matx_tensor_ops.h (hunk 5 above)
Comment (luitjens) [reply include/matx_tensor_ops.h]: fixed

File: include/matx_conv.h (hunk 4 above)
Comment (luitjens) [reply include/matx_conv.h:213]: fixed

File: include/matx_conv.h (hunk 1 above)
Comment (luitjens) [reply include/matx_conv.h]: fixed

**Pull request merged**